handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,203 @@
+     * @param closeOnTimeout if channel should be closed by this handler after firing {@link ReadTimeoutException}
+     *
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true);
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        init(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelFutureListener));
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx channel handler context of related {@link #channelRead} invocation
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
+        return true;
+    }
+
+    private void init(final ChannelHandlerContext ctx) {
+        removed = false;
+        if (channelFutureListener == null) {
+            channelFutureListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    if (future.isSuccess()) {
+                        initialize(ctx);
+                    }
+                }
+            };
+        }
+        if (readTimeoutTask == null) {
+            readTimeoutTask = new ReadTimeoutTask(ctx);
+        }
+    }
+
+    private void initialize(ChannelHandlerContext ctx) {
+        if (removed) {
+            return;
+        }
+
+        writeFinishTime = System.nanoTime();
+        readFinishTime = writeFinishTime - 1;
```

Comment:
Yes, if we don't have response - writeFinishTime - readFinishTime > 0 would be true after this initialization.